enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,348 @@
+through [Authorization Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3) flow
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified to the following steps:
+
+1. User runs `oc login --auto $CLUSTER_API_URL`
+2. `oc` starts a webserver to receive a callback from the OAuth server. It then opens the authorization endpoint with a
+   PKCE code challenge.
+3. The OAuth application performs the authentication either directly or by redirecting the user.
+4. Once authentication is successful the OAuth application redirects the user back to the callback URL of `oc` along
+   with a code.
+5. `oc` can then use this code along with the code challenge to request an access token. On receiving the token `oc`
+   then performs any configuration required to use the token.
+
+Using the above flow the user has to perform a login only once and the `oc` tool is configured automatically without any
+user input.
+
+### Goals
+
+Provide a one-step login for users which can be specified by the cluster operator.
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+### User Stories
+
+TBD
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have new 2 new flags, `--auto` and `--oauth-url`. The first flag indicates to `oc` that it
+should use the Oauth2 Authorization Flow to login. The second flag is optional and should be specified when the domain
+of the internal OAuth server has
+been [customized](https://docs.openshift.com/container-platform/4.9/authentication/configuring-internal-oauth.html#customizing-the-oauth-server-url_configuring-internal-oauth)
+. It will be the URL of the modified OAuth server.
+
+#### OAuth server
+
+The internal OAuth server uses [OSIN](https://github.com/openshift/osin) as a dependency to provide OAuth2 Authorized
+Grant Flow for a registered client to fetch tokens. It also has support for PKCE(Proof Key for Code Exchange) as
+specified in [rfc7636](https://datatracker.ietf.org/doc/html/rfc7636). _rfc7636_ is an enhancement for public clients,
+i.e. clients which cannot hold client credentials to fetch an access token securely. The new `oc login` is also a public
+client because it cannot contain client credentials to retrieve the access token and PKCE is designed for use cases like
+this one.
+
+Another change which may be required is to allow the port on the callback URL to be variable when the callback is to
```

Comment:
I'd pick any free port, you're supplying the callback as part of the URL including port anyway. The question (similar to the previous one above) is whether we need to explicitly whitelist the localhost domain - or in addition also have the port.

Reply:
azure has port 50342 btw: https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/profile/__init__.py#L56

Reply:
they also don't really make use of that port funnily enough since they switched to msal: https://github.com/Azure/azure-cli/blob/c6afbd09efe35cebd72beeb47ba6d1893d28434e/src/azure-cli-core/azure/cli/core/auth/identity.py#L119-L123

Reply:
Any free port outside the of the reserved ports range, you don't want to have to run `oc` with root privileges just because of the port choice.

What happens on Windows and Mac OS? I hear Mac OS might attempt to prevent you from opening localhost ports.

Reply:
I would make we wonder if IdP allow dynamic range of redirect ports, do they?

Reply:
> I would make we wonder if IdP allow dynamic range of redirect ports, do they?
It doesn't currently. However the server can be modified to ignore the ports if the callback is to `localhost` or `127.0.0.1`. I don't think the implications of this would be severe since there's also PKCE. So even if there is a hijack the attacker would not be able to do anything with the temp token.

Reply:
> Do you know anyone that has a mac around here to attest?

I do. Reach me on #forum-apiserver on Slack, we may be able to figure someone out.

Reply:
I've tested this on Windows and when started from the command line in Windows an application can listen even on ports like 8080.